File: chapter_004/rpn.py
#!/bin/bash/env python3

stack = []

s = input('RPN > ')
tokens = s.split()
for token in tokens:
    token = token.strip()
    try:
        stack.append(float(token))
    except ValueError:
        arg2 = stack.pop()
        arg1 = stack.pop()
        if token == '+':
            stack.append(arg1 + arg2)
        elif token == '-':
            stack.append(arg1 - arg2)
        elif token == '*':
            stack.append(arg1 * arg2)
        elif token == '/':
            stack.append(arg1 / arg2)
        else:
            print('Illegal operator: {}'.format(token))
            exit(1)

print(stack.pop())

File: chapter_004/rpn.rs
use std::io;

fn main() {
    let mut stack: Vec<f64> = vec![];
    loop {
        print!("RPN > ");
        let mut input = String::new();
        io::stdin().read_line(&mut input).expect("Failed to read line");
        let input = input.split_whitespace();
        for token in input {
            let token = token.trim();
            match token.parse::<f64>() {
                Ok(v) => { stack.push(v); continue; },
                Err(_) => 0.0,
            };

            let b = stack.pop().expect("Stack underflow");
            let a = stack.pop().expect("Stack underflow");
            match token {
                "+" => stack.push(a + b),
                "-" => stack.push(a - b),
                "*" => stack.push(a * b),
                "/" => stack.push(a / b),
                "%" => stack.push(a % b),
                "^" => stack.push(a.powf(b)),
                _ => println!("Unknown operator: {}", token),
            }
        }

    println!("{}", stack.pop().unwrap());
    }
}

